server-base/src/main/java/org/apache/kylin/rest/service/QueryService.java

```diff
@@ -591,21 +610,52 @@ private String getUserName() {
     }
 
     public SQLResponse searchQueryInCache(SQLRequest sqlRequest) {
-        SQLResponse response = null;
-        Cache exceptionCache = cacheManager.getCache(EXCEPTION_QUERY_CACHE);
-        Cache successCache = cacheManager.getCache(SUCCESS_QUERY_CACHE);
-
-        Element element = null;
-        if ((element = exceptionCache.get(sqlRequest.getCacheKey())) != null) {
-            logger.info("The sqlResponse is found in EXCEPTION_QUERY_CACHE");
-            response = (SQLResponse) element.getObjectValue();
+        Cache queryCache = cacheManager.getCache(QUERY_CACHE);
+        Cache.ValueWrapper wrapper = queryCache.get(sqlRequest.getCacheKey());
+        if (wrapper == null) {
+            return null;
+        }
+        SQLResponse response = (SQLResponse) wrapper.get();
+        if (response == null) {
+            return null;
+        }
+
+        // Check whether duplicate query exists
+        while (response.isRunning()) {
+            // Wait at most one minute
+            if (System.currentTimeMillis() - response.getDummyTime() >= 60000L) {
```

Comment:
better to make 60000L configurable?

One more question: can this dummy feature handle pushdown queries?

Reply:
This feature can handle pushdown queries.